packages/components/src/List/List.tsx

```diff
@@ -2,44 +2,91 @@ import React from "react";
 import get from "lodash/get";
 import groupBy from "lodash/groupBy";
-import styles from "./List.module.css";
+import { Typography } from "@jobber/components/Typography";
+import {
+  BaseListItemProps,
+  ListItem,
+  ListItemProps,
+} from "@jobber/components/List/ListItem";
 import sectionStyles from "./SectionHeader.module.css";
-import { ListItem, ListItemProps } from "./ListItem";
-import { Typography } from "../Typography";
+import styles from "./List.module.css";
 
-interface ListProps {
+interface ListProps<T extends BaseListItemProps = ListItemProps> {
   /**
    * Array of the list items.
    */
-  readonly items: ListItemProps[];
+  readonly items: T[];
+  /**
+   * A function that will be called for each item instead of the default
+   * rendering
+   */
+  readonly customRenderItem?: (item: T) => React.ReactNode;
+  /**
+   * Set to true for more control over the item styles
+   */
+  readonly customItemStyles?: boolean;
 }
 
-export function List({ items }: ListProps) {
-  const isSectioned = items.some(item => item.section);
+export function List<T extends BaseListItemProps = ListItemProps>({
+  items,
+  customRenderItem,
+  customItemStyles,
+}: ListProps<T>) {
+  const isSectioned = items.some(item => "section" in item && item.section);
 
   if (isSectioned) {
-    return <SectionedList items={items} />;
+    return (
+      <SectionedList
+        items={items}
+        customRenderItem={customRenderItem}
+        customItemStyles={customItemStyles}
+      />
+    );
   } else {
-    return <DisplayList items={items} />;
+    return (
+      <DisplayList
+        items={items}
+        customRenderItem={customRenderItem}
+        customItemStyles={customItemStyles}
+      />
+    );
   }
 }
 
-function DisplayList({ items }: ListProps) {
+function DisplayList<T extends BaseListItemProps = ListItemProps>({
+  items,
+  customRenderItem,
+  customItemStyles,
+}: ListProps<T>) {
+  const omitDefaultStyles = customRenderItem && customItemStyles;
+  const itemClassNames = classnames(!omitDefaultStyles && styles.item);
+
   return (
     <ul className={styles.list}>
       {items.map(item => (
-        <li key={item.id} className={styles.item}>
-          <ListItem {...item} />
+        <li key={item.id} className={itemClassNames}>
+          <ListItem
+            {...item}
+            customRenderItem={customRenderItem}
+            customItemStyles={customItemStyles}
+          />
         </li>
       ))}
     </ul>
   );
 }
 
-function SectionedList({ items }: ListProps) {
+function SectionedList<T extends BaseListItemProps = ListItemProps>({
+  items,
+  customRenderItem,
+  customItemStyles,
+}: ListProps<T>) {
   const sectionedItems = groupBy(items, item => get(item, "section", "Other"));
   const sectionHeaderClassNames = classnames(sectionStyles.sectionHeader);
 
+  const omitDefaultStyles = customRenderItem && customItemStyles;
+  const itemClassNames = classnames(!omitDefaultStyles && styles.item);
```

Comment:
I'm tempted to reduce this repetition since we're doing the exact same thing in `DisplayList`. though, it's 2 place, which isn't that bad at all.... plus what, it would save us 1 line of code in each component.

ehh. it's probably fine. I don't have a very compelling argument to do it right now.

Reply:
I thought about this as well, I landed on the fact that it would be preferable to have the interface between the components be simpler (since they are already so complicated) and live with the duplicated line